source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,467 @@
+                tracing::trace!("CMD2 response: {value:?}");
+                // TODO: map [u32; 4] value to u128
+                Ok(CardIdentification(0))
+            }
+        }
+    }
+
+    /// Get the relative card address
+    pub async fn get_rca(&mut self) -> Result<u32, Error> {
+        match self
+            .cmd(Command {
+                index: 3,
+                argument: 0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD3 response: {value:#x}");
+                Ok(value)
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+    }
+
+    /// Toggle the card between stand-by and transfer state
+    pub async fn select(&mut self, rca: u32) -> Result<CardStatus, Error> {
+        match self
+            .cmd(Command {
+                index: 7,
+                argument: rca,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD7 response: {value:#x}");
+                Ok(CardStatus(value))
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+    }
+
+    /// Use 4 data lanes
+    pub async fn set_wide_bus(&mut self, rca: u32) -> Result<CardStatus, Error> {
+        // Go to *APP* mode before sending application command
+        self.cmd(Command {
+            index: 55,
+            argument: rca,
+            options: HardwareOptions::None,
+            kind: CommandKind::Control,
+            rsp_type: ResponseType::Short,
+            rsp_crc: true,
+            buffer: None,
+        })
+        .await?;
+
+        match self
+            .cmd(Command {
+                index: 6,
+                argument: 0b10,
+                options: HardwareOptions::SetBusWidth(BusWidth::Quad),
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("ACMD6 response: {value:#x}");
+                Ok(CardStatus(value))
+            }
+            Response::Long(_) => Err(Error::Response),
+        }
+    }
+
+    /// Read the desired number of data blocks into the provided buffer, starting at the given sector.
+    pub async fn read(
+        &mut self,
+        sector: u32,
+        blocks: usize,
+        buf: FixedVec<u8>,
+    ) -> Result<FixedVec<u8>, Error> {
```

Comment:
i think we might want to validate that  `buf.capacity()` is >= `blocks * 512` here and return an error if it isn't? the actual sdmmc service implementation does check that before configuring the DMA descriptors, but it seems like it could be worthwhile to check here as well, so that we can avoid sending the request if the driver is just going to immediately reject it for not having a long enough buffer...

also, it might be worth having an additional layer of defense against buffer overruns if another platform implementation's SDMMC service doesn't correctly check before reading...